packages/peregrine/lib/talons/Accordion/useAccordion.js

```diff
@@ -0,0 +1,82 @@
+import {
+    Children,
+    cloneElement,
+    useCallback,
+    useEffect,
+    useState
+} from 'react';
+
+export const useAccordion = props => {
+    const { canOpenMultiple, children } = props;
+
+    const [openSectionIndicies, setOpenSectionIndicies] = useState(new Set([]));
+
+    const handleSectionClick = useCallback(
+        sectionIndex => {
+            const updatedIndicies = new Set(openSectionIndicies);
+
+            if (!openSectionIndicies.has(sectionIndex)) {
+                // The user wants to open this section.
+                if (!canOpenMultiple) {
+                    updatedIndicies.clear();
+                }
+
+                updatedIndicies.add(sectionIndex);
+            } else {
+                // The user wants to close this section.
+                updatedIndicies.delete(sectionIndex);
+            }
+
+            setOpenSectionIndicies(updatedIndicies);
+        },
+        [canOpenMultiple, openSectionIndicies, setOpenSectionIndicies]
+    );
+
+    // The accordion must tell each child section:
+    // 1. Whether it should be open or not.
+    // 2. What index it is.
+    // 3. A callback function to call when it gets clicked.
+    const controlledChildren = Children.map(children, (child, childIndex) => {
+        const isOpen = openSectionIndicies.has(childIndex);
+
+        return cloneElement(child, {
+            handleClick: handleSectionClick,
+            index: childIndex,
+            isOpen
+        });
```

Comment:
This is a clever solution, but `cloneElement` is generally an antipattern these days (barring any unusual circumstances), or at least a smell.

Why not use context to make `openSectionIndices` and a callback or two available to the child sections? That's the approach we've been using, for the most part.

Reply:
Sweet, thanks - done!